### ISSUE TITLE
LM should not block user enable module during deleting

### ISSUE THREAD

**Issue by ruanxin** [kyma-project/lifecycle-manager#836]
As reported by this [bug](https://github.com/kyma-project/lifecycle-manager/issues/833), LM will prevent the module from being upgraded if module CR is stuck under deleting. Which brings side effects.

The generic problem here is how we deal with the situation user disabled module, and module is under deletion, meanwhile, the user enabled it again.

The request here is not to block this user behaviour, and LM is still possible to hand status correctly.

AC:
- [ ] If module is under deletion, but user enables it during this time, module CR status should not be affected, other synced resources should be installed again.
- [ ] More specifically, if LM detects a new version to be applied (by channel switch, moduletemplate update), the upgrade process should not be blocked, which means all new resources should be installed.
- [ ] Manifest CR should marked for deleting (contains DeletionTimestamp) after all resources get deleted.

**Comment by jakobmoellerdev** [comment]
Interesting issue... From what I can sense there are a few things that are relevant here:
1. Module Enabled means a Spec, e.g. "desired State" of the module in the Kyma CR changed.
2. If the Module is enabled after it is deleted, the entire deletion needs to proceed as normal and the old Module needs to be removed entirely (check deletionTimestamp on Manifest for this, if timestamp is set then that means we have to wait), and after this we need to recreate everything again from scratch based on the ModuleTemplate.
3. If the LifecycleManager detects a new version while the deletion is progressing, it is simply not possible to change the image of the module deployment, simply because the deletionTimestamp stops specs upgrade completely. It is only possible to force remove the module by dropping finalizers and possibly killing state.
4. IMO it is bad practice to mark for deletion after resources are deleted. This is against kubernetes best practices and also does not respect deletionTimestampGracePeriod in k8s.

Basically this is a case of delete Module, wait until gone / suceeded or force removal, then recreate new module from spec.

Everything else is a reimagination of the kubernetes spec and should be discouraged.

**Comment by ruanxin** [comment]
In the ideal case, that's true, but in practice, we will have some challenges preventing module operator upgrades during deletion. For example, those security issues mentioned [here](https://github.com/kyma-project/lifecycle-manager/issues/833#issuecomment-1697009603) will be a problem if the module operator is not up-to-date for a long time.

Regarding the idea of postponing deleting manifest CR, it's basically trying to keep manifest CR not violating the k8s principle. If we immediately delete manifest CR (set deletionTimestamp), then there is no chance of bringing it to a normal state before deleting it completely.

The simplified situation here is how to manage manifest CR, module CR, and module operator consistently and comply with the following rules:
- module operator should exist if module CR exists (even under deleting) since it manages the lifecycle of module CR, e.g: create or remove finalizer.
- module operator should be up-to-date if exists.
- manifest CR can't be deleted (completely gone) before module operator and module CR are removed, otherwise LM can't manage related module resources, as well as propagate states. (This might be arguable)

At the moment, it's hard to imagine a better alternative solution. For example, if the user enables module again, we allow LM force delete manifest CR and recreate a new one, but then the existing status of the module should also be transferred correctly to this new manifest CR (have more edge cases needs to care).

**Comment by jakobmoellerdev** [comment]
> In the ideal case, that's true, but in practice, we will have some challenges preventing module operator upgrades during deletion. For example, those security issues mentioned [here](https://github.com/kyma-project/lifecycle-manager/issues/833#issuecomment-1697009603) will be a problem if the module operator is not up-to-date for a long time.

In a managed environment, if the BTP operator is outdated because its stuck in deleting, imo that is the operator fault. Although I must admit maybe I dont have the full picture since the lines got blurred with purge reconciler etc. I do have to say though that Lifecycle Manager should not be put into consideration if a module gets stuck during deletion and then doesnt get upgraded, for me that is an SRE scenario.

> Regarding the idea of postponing deleting manifest CR, it's basically trying to keep manifest CR not violating the k8s principle. If we immediately delete manifest CR (set deletionTimestamp), then there is no chance of bringing it to a normal state before deleting it completely.

Totally agree on this, but not setting deletionTimestamp is the equivalent of not deleting in kubernetes, its basically working around the way that kubernetes was designed, bad design imo. I think we can introduce deletionGracePeriod and force remove the module (just like the purge) and then reintroduce a new version on recreation. That would also fix security issues.

> The simplified situation here is how to manage manifest CR, module CR, and module operator consistently and comply with the following rules:
> 
> * module operator should exist if module CR exists (even under deleting) since it manages the lifecycle of module CR, e.g: create or remove finalizer.

IMO this is always the case as long as we make sure to not remove the Manifest CR but leave a finalizer in it. If the module operator itself needs to do cleanup logic, then it should also use finalizers. We need to respect that we cannot remove the operator while there are still finalizers on the module CR, maybe thats a good criteria to look for before removing the manager.

> * module operator should be up-to-date if exists.

IMO, The module operator should be on the desired state of the module BEFORE it was triggered for deletion. That way the delete should progress normally (or get stuck and get resolved by SRE or purge) and then get recreated with up-to-date information

> * manifest CR can't be deleted (completely gone) before module operator and module CR are removed, otherwise LM can't manage related module resources, as well as propagate states. (This might be arguable)

Completely agree on this, see my above point to how I think we could solve this.

> At the moment, it's hard to imagine a better alternative solution. For example, if the user enables module again, we allow LM force delete manifest CR and recreate a new one, but then the existing status of the module should also be transferred correctly to this new manifest CR (have more edge cases needs to care).

Why do you allow force deletion of a manifest? Shouldn't the module wait for proper cleanup, then recreate itself? In which case where Module is not faulty and needs manual intervention anyhow do we need force-deletion?

**Comment by ruanxin** [comment]
So in summary, what you suggested is to separate the dependencies between Module Operator (with other resources) and Module CR **during deleting**, say configure 5 mins with `terminationGracePeriodSeconds`, so that the operator still has enough time to deal with clean up managed resources. 

Also, removing manifest CR (remove finalizer so that it can be deleted) should not determined by the existence of Module CR, basically the manifest reconciler stuck at [preDelete](https://github.com/kyma-project/lifecycle-manager/blob/main/internal/declarative/v2/reconciler.go#L474-L476) step during deleting, but only check if other resources exists.  

One thing to notice is, atm, if we remove the preDelete check, then the manifest reconciler will not check other resources, but just do delete request once, then remove its own finalizer immediately.  

Is above mentioned aligned with your thoughts @jakobmoellerdev 

Since you mentioned purge reconciler, just want to remark here, it's not involved purge reconciler, which only works when kyma CR is under deleting, but the scenario we are talking about is only triggered by users, and the SKR shoot always exists.


>Totally agree on this, but not setting deletionTimestamp is the equivalent of not deleting in kubernetes, its basically working around the way that kubernetes was designed, bad design IMO. 
I'm not saying not set `deletionTimestamp`, but just postpone the time LM sets it, atm, the delete request sent by LM is when user removes module in `kyma.spec` by kyma reconciler, but what I'm thinking here is postpone the set when LM detect there is no sync resources exists, which managed by manifest reconciler.


> Why do you allow force deletion of a manifest? Shouldn't the module wait for proper cleanup, then recreate itself? In which case where Module is not faulty and needs manual intervention anyhow do we need force-deletion?
No, that's just an example as a bad alternative solution, to illustrate the side effects it will bring.

**Comment by janmedrek** [comment]
@jakobmoellerdev the problematic part is the requirement that we need to somehow "protect" the workload of customers from themselves.

For example - if the user creates a serverless Function CR, then the whole infrastructure part supporting that Function CR gets created by the Serverless Operator. The Function runs properly. Later, the Serverless module gets disabled in the cluster, this is where the requirement comes in - the finalizers are set and the Function should still work until the user removes those Function CRs.

There are similar cases for other modules (BTP Manager included), it ends up in a situation where the user is stuck in a "Deleting" state despite no errors either on the LM or operator side.

Regardless of whether the requirement makes sense or not - it would mean that keeping the module in a "Deleting" state is a feature that we somehow need to support.

As I already wrote [here](https://github.com/kyma-project/lifecycle-manager/issues/833#issuecomment-1721028253), my proposal would be to keep the deletion logic separated from the update logic, then the deletion process would be split into two phases:
1. Remove CR (set deletion timestamp, wait for the finalizer to be gone, wait for the CR to be gone)
2. Delete all the rest of the resources listed in the manifest

This way in phase 1 we could still keep updating all the resources if a new ModuleTemplate is applied. All the issues in phase 2 would be explicitly handled by the operator team as any errors there are related only to the technicalities of modules/LM.

**Comment by jakobmoellerdev** [comment]
The way to solve this is by having the module operators take care of their domain. When lifecycle manager resources get touched, lifecycle manager should clean up the state that it manages. If the CR is removed, then the module operator needs to know if it is safe to keep the rest of the resources in there before dropping its finalizer.

The only reason we have modules stuck in deletion is because the module operator itself is faulty during deletion. Faulty means in this case that they should not be blocking their deletion if there are still CRs in the cluster. 

The ModuleTemplate the module operator provides should not even contain a default CR if that CR should not be cleaned up by Lifecycle Manager on a delete. Instead it should be provisioned by the Module Operator on startup (check if exists, if not create). For a delete, this means that Lifecycle Manager then does not care about the Module CR. That way nothing can get stuck and the operator gets removed and recreated.

**Comment by jakobmoellerdev** [comment]
On a side note: We do not touch the operator deployment until the CR is gone (if it is included in the ModuleTemplate and managed by us). 

That means that all Functions will continue to work until the module operator does some cleanup itself. If it does not do that cleanup when the CR is triggered for deletion, the module operator is buggy.

**Comment by ruanxin** [comment]
@janmedrek @jakobmoellerdev , 

Let's first align with the current implementation:

The deletion logic is already implemented as required, it prevents resources to deleted until moduleCR is gone, the [preDelete](https://github.com/kyma-project/lifecycle-manager/blob/main/internal/manifest/custom_resource.go#L89) will return error if moduleCR exists.

Why it blocks module to be updated?
In [this](https://github.com/kyma-project/lifecycle-manager/blob/main/internal/declarative/v2/reconciler.go#L410-L416) logic, if LM detected manifestCR is under deleting, the target resources (contains all new resources)will be an empty list to prevent update happens.

Actually, if we change the above logic when returning empty target resources to `if [manifestCR has deletingTimeStamp] && [moduleCR does not exists]`, then this could fulfil the requirement I believe.


@jakobmoellerdev, the current problem to prevents us from having this ideal deleting behaviour is mainly we can't make sure each module is designed and implemented as a "real module", we don't know (maybe don't want to think about it in detail) if deleted resources contain something the user doesn't want to delete (intentionally or by mistake). IMO, if we can distinguish the ownership of resources, means which resources owned by users (can be ignroed during delete), which resources owned by operators (can be safely marked with deletingtimestamp), then LM are possible to deal with deleting logic correctly.

**Comment by jakobmoellerdev** [comment]
I can only emphasize again. You are breaking kubernetes standards by continuing background propagation of an owned resource by ignoring the fact that the resource still exists. If the CR is then still existing and the module is removed, then the next time the module is created the CR will still have the deletionTimestamp and will block creation of the Module CR because the old CR with deletionTimestamp still exists.

The only way to solve this is by having Module Operators not use the inbuilt CR that is managed by lifecycle manager and instead have them manage it on their own.

**Comment by ruanxin** [comment]
> I can only emphasize again. You are breaking kubernetes standards by continuing background propagation of an owned resource by ignoring the fact that the resource still exists. If the CR is then still existing and the module is removed, then the next time the module is created the CR will still have the deletionTimestamp and will block creation of the Module CR because the old CR with deletionTimestamp still exists.

Yes, this is clear, if user enabled a module under deleting, the state of this module will not change, still `deleting`, the manifest CR and module CR both have deletionTimestamp. It will keep this state util user manually removed module CR, then all the rest resources will be deleted and recreated, then this module back to normal `ready` state.

**Comment by jeremyharisch** [comment]
**ACs**

- [ ] Change [this logic](https://github.com/kyma-project/lifecycle-manager/blob/main/internal/declarative/v2/reconciler.go#L410-L416) when returning empty target resources to if `[manifestCR has deletingTimeStamp] && [moduleCR does not exists]`
- [ ] Don't return error in the [preDelete](https://github.com/kyma-project/lifecycle-manager/blob/main/internal/manifest/custom_resource.go#L89) if moduleCR is still exists during deleting to unblock followup logic.
- [ ] Cover with e2e tests:
   - Scenario 1: Test module CR still possible to be updated during deleting state
   - Scenario 2: Test if module is set to `Ready` state if module is re-enabled, after module CR has been deleted
   - Use template-operator as module under test
